pykechain/models/base_reference.py

```diff
@@ -0,0 +1,129 @@
+from abc import abstractmethod, ABC
+from typing import Optional, Any, Text, Iterable, Union, Tuple, List
+
+from pykechain.models import Property2, Base
+from pykechain.models.base import BaseInScope
+from pykechain.utils import is_uuid
+
+
+class _ReferenceProperty(Property2):
+    """
+    Private base class for the KE-chain reference properties.
+
+    .. versionadded:: 3.7
+    :cvar REFERENCED_CLASS: handle to the Pykechain class that is referenced using this ReferenceProperty class.
+    """
+
+    REFERENCED_CLASS = Base  # type: type(Base)
+
+    def __init__(self, json, **kwargs):
+        """Construct a ReferenceProperty from a json object."""
+        super().__init__(json, **kwargs)
+
+        self._cached_values = None
+
+    @property
+    def ref_name(self) -> Text:
+        """Get the name of the referenced Pykechain class."""
+        return self.REFERENCED_CLASS.__name__
+
+    @property
+    def value(self) -> Optional[Iterable[REFERENCED_CLASS]]:
+        """
+        Retrieve the referenced objects of this reference property.
+
+        :return: list or generator of `Base` objects.
+        :rtype list
+        """
+        if not self._value:
+            return None
+        elif not self._cached_values:
+            self._cached_values = self._retrieve_objects()
+        return self._cached_values
+
+    @value.setter
+    def value(self, value: Any) -> None:
+        if self.use_bulk_update:
+            self._pend_value(value)
+        else:
+            self._put_value(value)
+
+    @abstractmethod
+    def _retrieve_objects(self, **kwargs) -> Iterable[Base]:
+        """
+        Retrieve a list of Pykechain objects, type depending on the reference property type.
+
+        This method is abstract because the exact method of the Client class changes per subclass.
+
+        :param kwargs: optional arguments
+        :return: list of Pykechain objects inheriting from Base
+        """
+        pass
+
+    def serialize_value(self, value: Union[List, Tuple]):
```

Comment:
What is the output type? I guess it is `-> Optional[List[ObjectID]]`

Reply:
Correct, so i'll use `Optional[List[Text]]`